Use a separate ExecutorService to fix #71.

bturner opened the pull request:
- Updated AsyncProcessor to look at the Bitbucket Server verson and choose between using the shared ExecutorService (5.9+) or creating its own ExecutorService (previous versions)
  - Bitbucket Server 5.9 includes a fix which prevents add-ons using the shared BucketedExecutor from blocking hook callbacks
- Removed some dead code in AsyncProcessor and PullRequestListener
- Simplified Maven structure and reconfigured how integration and unit tests are distinguished
  - Moved unit tests from src/test/java/ut to src/test/java
  - Added a simple UserResourceTest which requests configuration for a repository which hasn't had any applied
  - This verifies that the PR Harmony plugin has been installed and enabled successfully, and that its REST resource is available
- Simplified some of the processing in UserUtils to use the API more efficiently and avoid various IDEA inspection warnings

bturner commented on pom.xml:
```diff
@@ -33,64 +33,79 @@
     <dependency>
       <groupId>com.atlassian.templaterenderer</groupId>
       <artifactId>atlassian-template-renderer-api</artifactId>
-      <version>1.5.7</version>
```
In general, if you have a `<scope>import</scope>` for `bitbucket-parent`, you shouldn't define explicit versions on your dependencies. Our parent POM defines the versions of the libraries that we actually ship with, so using those versions ensures you're (for example) using the "right" APIs for Bitbucket Server versions you target.

monitorjbl replied:
Thanks for the tip. The POM was largely copied from Atlassian's example plugin project, so if you haven't already you might want to update that to prevent others from doing this.

bturner commented on pom.xml:
```diff
@@ -33,64 +33,79 @@
     </dependency>
-    <dependency>
-      <groupId>commons-lang</groupId>
-      <artifactId>commons-lang</artifactId>
```
This was unused

bturner commented on pom.xml:
```diff
@@ -99,47 +114,9 @@
       <scope>test</scope>
     </dependency>
-
-    <!-- WIRED TEST RUNNER DEPENDENCIES -->
```
I've replaced all of these dependencies with new dependencies on Bitbucket Server's common test modules (`bitbucket-it-common` and `bitbucket-page-objects`). Those will transitively pull in all the dependencies you need to write browser and REST integration tests.

bturner commented on pom.xml:
```diff
@@ -99,47 +114,9 @@
   </dependencies>
 
   <build>
-    <testSourceDirectory>${project.basedir}/src/test/java/ut</testSourceDirectory>
```
Rather than separate the source like this, I've switch the setup to have Surefire (used to run unit tests) ignore everything under `src/test/java/it`, and have AMPS (which uses Failsafe to run integration tests) ignore everything that's _not_ under `src/test/java/it`.

(Feel free to not do it this way; this is just how we write our own plugins.)

bturner commented on pom.xml:
```diff
@@ -151,64 +128,55 @@
+    <bitbucket.data.version>${bitbucket.version}</bitbucket.data.version>
+    <!-- This separate version allows testing the plugin with a different version of the system than
+         it's compiled against, which is useful for testing compatibility with multiple versions -->
+    <bitbucket.test.version>${bitbucket.version}</bitbucket.test.version>
```
I'd introduced this separate version to allow, for example, compiling PR Harmony against Bitbucket Server 4.8 (which is the lower bound of the compatibility range you've set on Marketplace), but running the tests against Bitbucket Server 5.8 (the upper bound).

You can run the integration tests against a given version with `mvn clean install -DskipITs=false -Dbitbucket.test.version=5.8.1`

monitorjbl replied:
Nice!

bturner commented on src/main/java/com/monitorjbl/plugins/AsyncProcessor.java:
```diff
@@ -1,32 +1,48 @@
-public class AsyncProcessor {
-  private static final String PREFIX = "com.monitorjbl.plugins:pr-harmony:";
-  private final ExecutorService concurrencyService;
+public class AsyncProcessor implements LifecycleAware {
```
`LifecycleAware` means Bitbucket Server will automatically trigger `onStart` and `onStop` at appropriate points. For Bitbucket Server 5.8 and older, this allows `onStop` to shutdown the PR Harmony-managed `ExecutorService`.

bturner commented on src/main/java/com/monitorjbl/plugins/AsyncProcessor.java:
```diff
@@ -1,32 +1,48 @@
+    } else {
+      //Bitbucket Server 5.9 includes a fix for BSERV-10652, which moves hook callback to their own threadpool, so
+      //for 5.9+ use the shared executor
+      this.executorService = executorService;
```
For 5.9, this allows PR Harmony to continue using our shared `ExecutorService`. You don't have to do it this way (you could _always_ use your own), but I'd prefer (and appreciate!) it if you did. This way the background processing PR Harmony triggers "shares" the server with other background processing, rather than always running in addition to it. That should help manage overall server load.

monitorjbl replied:
Will this not cause thread exhaustion when the server gets a lot of build updates? If the executor wasn't shared, it wouldn't directly interfere with that (although it certainly could interfere with CPU/memory consumption).

bturner replied:
This is a (difficult) balancing act.

Certainly, having a significant number of build status changes could result in PR Harmony creating a lot of tasks for the shared executor. In a system where there are many add-ons using the shared executor, those build status tasks may delay tasks created by other add-ons--just like a large backlog of tasks from other add-ons would be able to delay PR Harmony's build status processing.

For Bitbucket Server, though, that's actually _desirable_, as long as the tasks being delayed are not critical. That's the real "bug" in BSERV-10652; we were using the shared executor for _critical_ processing: Hook callbacks. Having a shared limit on background processing that can be delayed helps to ensure the accumulation of add-ons doesn't overwhelm the system.

(That's not to say that PR Harmony's build status processing is not _important_; clearly it is. But a delay of minutes is less critical than with, say, hosting operations.)

Now, it could certainly be contended that if PR Harmony just always used its own executor, it wouldn't starve critical processing, but it also wouldn't starve, or be starved by, other add-ons. From a purely PR Harmony standpoint, that's all good. But from _my_ standpoint, which I'll readily admit is biased, it's less good. The build status processing PR Harmony does is quite heavy, and produces non-trivial system load. When load becomes too high, it's usually not add-on vendors that see the support load, at least initially; it's Bitbucket Server's developers and supporters.

If all add-ons create and use their own executors, it's likely some (most? all?) of them will use non-configurable sizes (as I've done here, I'll note, for 5.8 and earlier). But even if they _are_ configurable, that still means a system administrator has to find out what all those configuration properties are and apply them--and they'd need to have some way of readily determining which add-on(s) executor(s) they need to reconfigure based on who's starving (or who's creating too much load). That's not always straightforward to figure out, especially for non-programmers.

By comparison, if add-ons use the shared executor, it's a single thing to size. That implies, if tasks are getting backed up, administrators can consider their current system load and opt to add more threads to the shared executor, which would then be "spread" across all the add-ons that use it and allow the backlog to be cleared more quickly. Additionally, we're planning to add monitoring for the queue size on the shared executor in a future release of Bitbucket Server, which will make it easier for administrators to track the size of the backlog on their server. That's another feature add-ons which use their own executors won't benefit from. We may add an SPI that they could use (not sure what the plan is there, at the moment), but they'd still have to manually build it out.

bturner commented on src/main/java/com/monitorjbl/plugins/PullRequestListener.java:
```diff
@@ -20,9 +20,6 @@
 public class PullRequestListener {
-  private static final String PR_APPROVE_BUCKET = "AUTOMERGE_PR_APPROVAL";
-  private static final String BUILD_APPROVE_BUCKET = "AUTOMERGE_BUILD_APPROVAL";
-  public static final int MAX_COMMITS = 1048576;
```
All of these were unused, with the switch to streaming and moving away from the `BucketedExecutor`

bturner commented on src/main/java/com/monitorjbl/plugins/UserUtils.java:
```diff
@@ -19,44 +23,44 @@ public UserUtils(UserService userService) {
+      return Collections.emptyList();
+    }
+
+    Set<ApplicationUser> usersByName = userService.getUsersByName(distinctNames);
```
Using `getUsersByName` here allows the server to perform a significantly more efficient lookup, retrieving users with a single database query instead of one per user.

monitorjbl replied:
What happens if there are a lot of users? Is this paged?

bturner replied:
No. You've always been accumulating all the slugs into a single `List`, though, so the end result is identical memory usage. The only place this will be different is it can have N `ApplicationUser` instances in memory while assembling that list instead of just one at a time.

"A lot" translates to a pretty big number, especially for something this short-lived, though, and users are fairly small. Is it common for people to configure 10,000 default or required reviewers? This is something of a trade-off between optimizing the common use case, which I'd imagine has 10-20 users tops, against optimizing memory usage in edge cases.

bturner commented on src/main/java/com/monitorjbl/plugins/UserUtils.java:
```diff
@@ -19,44 +23,44 @@ public UserUtils(UserService userService) {
   public String getUserDisplayNameByName(String username) {
-	    ApplicationUser user = userService.getUserByName(username);  
-	    return user.getDisplayName() == null ? username : user.getDisplayName();
+    return ofNullable(userService.getUserByName(username))
```
This change guards against `getUserByName` returning `null` (It's marked `@Nullable` in our documentation)

bturner replied:
Actually, from looking at the open issues for the plugin, this change will fix #69 (Or, at least, will prevent the `NullPointerException` from that issue from occurring)

bturner commented on src/main/java/com/monitorjbl/plugins/UserUtils.java:
```diff
@@ -19,44 +23,44 @@ public UserUtils(UserService userService) {
-        users.addAll(results.stream()
-            .map(ApplicationUser::getSlug)
-            .collect(Collectors.toList()));
-        results = newArrayList(userService.findUsersByGroup(group, new PageRequestImpl(i * RESULTS_PER_REQUEST, RESULTS_PER_REQUEST)).getValues());
```
So the code was requesting 25 users at a time. If 24 were returned (meaning there wasn't a 25th, and indicating there are no more users), because `results.size()` would still be `>0`, you'd do an extra page load trying to go "beyond" the end. When _that_ returned empty, you'd stop.

bturner commented on src/main/java/com/monitorjbl/plugins/UserUtils.java:
```diff
@@ -19,44 +23,44 @@ public UserUtils(UserService userService) {
+    return groups.stream()
+        //Replace each group with a stream of the users in that group
+        .flatMap(group ->
+            streamIterable(new PagedIterable<>(
```
Those code switches to `PagedIterable`, which automatically applies correct pagination and concatenates the pages together, on demand, into a single `Iterable`.

bturner commented on src/main/java/com/monitorjbl/plugins/config/UserResource.java:
```diff
@@ -31,12 +30,12 @@ public Response get(@PathParam("projectKey") String projectKey, @PathParam("repo
     return Response.ok(ImmutableMap.of(
         "requiredReviews", config.getRequiredReviews() == null ? "" : config.getRequiredReviews(),
         "blockMergeIfPrNeedsWork", config.getBlockMergeIfPrNeedsWork() == null ? "" : config.getBlockMergeIfPrNeedsWork(),
-        "requiredReviewers", utils.dereferenceUsers(newArrayList(newHashSet(concat(
```
This code was loading the dereferencing the groups, which implicitly returns "dereferenced" users by name, and then dereferencing those users _again_ together with any explicitly configured users. I've changed it to not re-dereference users from dereferenced groups. Instead, it only dereferences explicitly configured users.

(Note that this code could, and still can, return duplicates, if a user was in a configured group _and_ configured by name.)

monitorjbl replied:
The HashSet was guarding against duplicates. Was there a reason you removed that part?

bturner replied:
Ah, of course. You were deduping on the _input_ side, before. The old code didn't have any deduping on the _output_, but since you eliminated them from the input it wouldn't matter.

I'll re-add the set, but around the user/group call instead of inside. That way group users still don't get dereferenced twice.

bturner replied:
Turns out I'd introduced a more subtle breakage than that, in that `dereferenceGroups` and `dereferenceUsers` return different types (Strings and `User`s, respectively). I've fixed that now too.

However, noticing _that_ led me to notice a different bug. Specifically, the `String`s returned by `dereferenceGroups` are user _slugs_, not user _names_. But in `CommitBlockerHook`, they're being compared to `UserProfile.getUsername()`. That means in any instance where users' names and slugs are not identical, configuring excluded groups will not actually work. So I fixed that too.

The _best_ fix for that is to replace references to SAL's `UserManager` with Bitbucket Server's `AuthenticationContext`, which returns an `ApplicationUser` directly (and allows access to its slug). Using `AuthenticationContext` also simplifies `ConfigServlet` and removes the need for `UserUtils.getApplicationUserByName`, so I removed that too.

bturner commented on start-test-server.sh:
```diff
@@ -1,3 +1,3 @@
 #!/bin/bash
 
-atlas-run -u 6.3.0 --jvmargs "-agentlib:jdwp=transport=dt_socket,server=y,suspend=n,address=5005"
+atlas-run -u 6.3.14 --jvmargs "-agentlib:jdwp=transport=dt_socket,server=y,suspend=n,address=5005"
```
Just updated to match the AMPS version upgrade I applied in your `pom.xml`.

monitorjbl replied:
Thanks!

Commit:
Use a separate ExecutorService to fix #71.

- Updated AsyncProcessor to look at the Bitbucket Server verson and
  choose between using the shared ExecutorService (5.9+) or creating
  its own ExecutorService (previous versions)
  - Bitbucket Server 5.9 includes a fix which prevents add-ons using
    the shared BucketedExecutor from blocking hook callbacks
- Removed some dead code in AsyncProcessor and PullRequestListener
- Simplified Maven structure and reconfigured how integration and unit
  tests are distinguished
  - Moved unit tests from src/test/java/ut to src/test/java
  - Added a simple UserResourceTest which requests configuration for
    a repository which hasn't had any applied
  - This verifies that the PR Harmony plugin has been installed and
    enabled successfully, and that its REST resource is available
- Simplified some of the processing in UserUtils to use the API more
  efficiently and avoid various IDEA inspection warnings

bturner commented:
I've added a single REST integration test, which effectively verifies that the plugin _starts_ and _enables_ without issue. That means my code changes are structurally correct. However, I don't really have a good setup for actually _testing_ the plugin. I apologize for the hassle, but you'll probably want to verify these changes yourself before you merge.

bturner commented:
Please hold off on this for a moment; I'm fixing some bugs I introduced (and adding more integration and unit tests as penance for doing so).

Commit:
Rework for pull request #72.

- Added more integration tests
  - ConfigResource and UserResource are both fully tested
  - CommitBlockerHook is now tested using real pushes, in addition to
    its existing unit tests
- Fixed deduplication for users in UserResource.get's response
  - Also fixed an unintentional change in the "shape" of the payload
    where group members would have been returned as simple strings
    instead of Users
- Fixed an issue in CommitBlockerHook where checking for users excluded
  by groups was comparing usernames against slugs
  - UserUtils.dereferenceGroups returns _slugs_, not usernames
  - Also fixed the hook so it doesn't apply its rules to tags
  - Revised how the hook applies its checks to avoid dereferencing
    users and groups on every iteration of the loop, even if the ref
    being tested doesn't match any blocking pattern
  - Revised the rejection wording to include all blocked branches when
    a single push attempts to update more than one
- Removed RegexUtils.formatBranchName and instead simplified callsites
  to use MinimalRef.getDisplayId, which already omits refs/heads/
- Updated ConfigServlet to use AuthenticationContext to retrieve the
  current user instead of the SAL UserManager
  - This simplifies applying permission checks later
- Updated component-imports in atlassian-plugin.xml to ensure every
  component the add-on uses is explicitly imported

bturner commented:
Okay, I've pushed all my updates. I've verified my changes against Bitbucket Server 4.8.6 and 5.8.0, the oldest and newest versions the add-on claims to support. All seems well. (However, my validation has been based on my _tests_, so some things I haven't changed, like pull request auto-merging, haven't been tested. It's likely you'll still want to do your own validation.)

Apologies for the size of this pull request. I've probably gone too far fixing up issues. To try and increase your confidence in accepting my changes, I've added some new unit tests as well as several integration tests. `ConfigResource` and `UserResource` now have full REST tests, and `CommitBlockerHook` has an integration test which configures blocking and then validates handling of actual pushes end-to-end.

If my changes are too broad, you're welcome to pick them apart and use what suits you. I've generally tried to stick with what I perceive to be your code style (two space indenting, etc), but I may not have gotten everything quite right.

There's still one fairly pervasive issue with how the plugin is written that I haven't really tried to fix, and that's the inconsistent use of usernames and user slugs. For example, your `User` object has a `slug` field, but what you're actually putting in that field is the user's _username_, not their slug. I fixed a username/slug mismatch in `CommitBlockerHook` (which really should be `PushBlockerHook`, since Bitbucket Server cannot block _commits_; those are made on developers' workstations and then _pushed_ to Bitbucket Server, at which point it can reject them), but there are almost certainly more. In many systems a user's username and slug are identical, but that's _not guaranteed_. For example, e-mail addresses (user@example.com) are valid _usernames_, but the slug for such a user would replace the @ with an _ (user_example.com). You might try creating a user like that and then doing some testing.

bturner commented on src/main/java/com/monitorjbl/plugins/CommitBlockerHook.java:
```diff
@@ -1,54 +1,84 @@
-      if(regexUtils.match(config.getBlockedCommits(), branch) && !excluded.contains(user.getUsername())) {
-        hookResponse.err().write("\n" +
-                "******************************\n" +
-                "*    !! Commit Rejected !!   *\n" +
```
You can, of course, change this back, if you prefer, but it's really inaccurate. Bitbucket Server can't reject commits. They're authored remotely. What this is blocking is the _push_, preventing the new commits from being applied. It's a subtle distinction, perhaps, but it's an important one.

bturner commented on src/main/java/com/monitorjbl/plugins/CommitBlockerHook.java:
```diff
@@ -1,54 +1,84 @@
+
+    //Otherwise, if the user is pushing to at least one branch which does not allow direct commits, check
+    //whether the they've been excluded from the restriction
+    ApplicationUser user = authenticationContext.getCurrentUser();
```
We need the full `ApplicationUser`, rather than a SAL `UserProfile`, because exclusion _groups_ need to be checked against the _slug_. `UserUtils.dereferenceGroups` returns a collection of _slugs_, not usernames. (And it did before I started making changes, just to be clear.)

bturner commented on src/main/java/com/monitorjbl/plugins/PullRequestApproval.java:
```diff
@@ -29,7 +28,7 @@ public boolean isPullRequestApproved(PullRequest pr) {
   }
 
-  public Set<String> missingRevieiwers(PullRequest pr) {
+  public Set<String> missingReviewers(PullRequest pr) {
```
These two methods were misspelled. (Sorry, personal quirk)

bturner commented on src/main/java/com/monitorjbl/plugins/RegexUtils.java:
```diff
@@ -16,8 +14,4 @@ public boolean match(List<String> regexes, String value) {
     return false;
   }
-
-  public String formatBranchName(String refspec) {
```
You should never need to do something like this, in any modern (4.0+) version of the system. Where `RefChange` in Stash 3.11 and prior only included ref IDs, Bitbucket Server 4.0 replaced that with a `MinimalRef`, which exposes both the ID (refs/heads/example) and the display ID (just example).

I've removed this method, and updated all the callers to just use `getDisplayId()` instead.

bturner commented on src/main/java/com/monitorjbl/plugins/UserUtils.java:
```diff
@@ -3,64 +3,93 @@
 
 public class UserUtils {
-  private static final int RESULTS_PER_REQUEST = 25;
+  private static final int RESULTS_PER_REQUEST = 50;
```
I've made your pages a little bigger to improve performance. Bitbucket Server itself uses 25 as the default for some page sizes, but goes as high as 100 for some things.

bturner commented on src/main/java/com/monitorjbl/plugins/UserUtils.java:
```diff
@@ -3,64 +3,93 @@
+        .collect(toSet());
   }
 
-  public List<String> dereferenceGroups(Collection<String> groups) {
```
I've changed this to return a `Set` because that optimizes `contains` checks

bturner commented on src/main/java/com/monitorjbl/plugins/UserUtils.java:
```diff
@@ -3,64 +3,93 @@
         .orElse(username);
   }
-  
-  public ApplicationUser getApplicationUserByName(String username) {
```
With the switch from SAL's `UserManager` to Bitbucket Server's `AuthenticationContext` (which is what our implementation for SAL's `UserManager` uses under the covers), you don't need to do this lookup anymore when checking permissions in `ConfigServlet`; you already have a full `ApplicationUser`

bturner commented on src/main/resources/atlassian-plugin.xml:
```diff
@@ -8,17 +8,19 @@
+  <component-import key="securityService" interface="com.atlassian.bitbucket.user.SecurityService"/>
+  <component-import key="templateRenderer"
+                    interface="com.atlassian.templaterenderer.velocity.one.six.VelocityTemplateRenderer"/>
+  <component-import key="userService" interface="com.atlassian.bitbucket.user.UserService"/>
```
I've adjusted your `component-import`s to ensure every thing the plugin uses is explicitly imported.

Because this is a transformed plugin (we turn it into an OSGi bundle on your behalf when it's installed), the system can automatically fix missing imports in some places. But it's better to just import all your dependencies explicitly, rather than relying on that.

bturner commented on src/test/java/com/monitorjbl/plugins/PullRequestListenerTest.java:
```diff
@@ -11,72 +10,68 @@
+  private RegexUtils regexUtils;
+  @Spy
+  @SuppressWarnings("unused") //Used by @InjectMocks
+  private SecurityService securityService = new DummySecurityService();
```
Using our `DummySecurityService` makes your test a little simpler; no need for `MockSecurityContext` and related setup

bturner commented on src/test/java/com/monitorjbl/plugins/config/ConfigServletTest.java:
```diff
@@ -5,88 +5,80 @@
   public void testFoundRepo() throws Exception {
-    sut.handleRequest("/plugins/servlet/pr-harmony/PRJ/repo1", "user1", response);
-    verify(response, times(1)).setStatus(HttpServletResponse.SC_OK);
-    verify(renderer, times(1)).render(anyString(), any(Map.class), any(Writer.class));
```
`verify` is implicitly `times(1)`
